src/main/java/cloud/fogbow/ras/core/plugins/interoperability/cloudstack/compute/v4_9/CloudStackComputePlugin.java

```diff
@@ -107,165 +107,195 @@ public String requestInstance(ComputeOrder computeOrder, CloudStackUser cloudUse
         List<GetAllDiskOfferingsResponse.DiskOffering> diskOfferings = diskOfferingsResponse.getDiskOfferings();
 
-        if (diskOfferings != null) {
+        if (diskOfferings != null && !diskOfferings.isEmpty()) {
             for (GetAllDiskOfferingsResponse.DiskOffering diskOffering : diskOfferings) {
                 if (diskOffering.getDiskSize() >= diskSize) {
                     return diskOffering;
                 }
             }
         }
 
-        return null;
+        throw new NoAvailableResourcesException(
+                Messages.Error.UNABLE_TO_COMPLETE_REQUEST_DISK_OFFERING_CLOUDSTACK);
     }
 
-    private GetAllDiskOfferingsResponse getDiskOfferings(CloudStackUser cloudUser) throws FogbowException {
-        GetAllDiskOfferingsRequest request = new GetAllDiskOfferingsRequest.Builder().build(this.cloudStackUrl);
-        CloudStackUrlUtil.sign(request.getUriBuilder(), cloudUser.getToken());
+    @NotNull
+    @VisibleForTesting
+    GetAllDiskOfferingsResponse getDiskOfferings(@NotNull final CloudStackUser cloudUser)
+            throws FogbowException {
+
+        GetAllDiskOfferingsRequest getAllDiskOfferingsRequest = new GetAllDiskOfferingsRequest.Builder()
+                .build(this.cloudStackUrl);
+        URIBuilder uriRequest = getAllDiskOfferingsRequest.getUriBuilder();
+        String token = cloudUser.getToken();
+        CloudStackUrlUtil.sign(uriRequest, token);
 
-        String jsonResponse = null;
         try {
-            jsonResponse = this.client.doGetRequest(request.getUriBuilder().toString(), cloudUser);
+            String jsonResponse = doGet(uriRequest.toString(), cloudUser);
+            return GetAllDiskOfferingsResponse.fromJson(jsonResponse);
         } catch (HttpResponseException e) {
             CloudStackHttpToFogbowExceptionMapper.map(e);
+            return null;
         }
+    }
 
-        GetAllDiskOfferingsResponse diskOfferingsResponse = GetAllDiskOfferingsResponse.fromJson(jsonResponse);
+    @NotNull
+    @VisibleForTesting
+    String normalizeNetworksID(@NotNull final ComputeOrder computeOrder) {
+        List<String> networks = new ArrayList<>();
+        networks.add(this.defaultNetworkId);
+        List<String> userDefinedNetworks = computeOrder.getNetworkIds();
+        if (!userDefinedNetworks.isEmpty()) {
+            networks.addAll(userDefinedNetworks);
+        }
+        return StringUtils.join(networks, ",");
+    }
 
-        return diskOfferingsResponse;
+    @NotNull
+    @VisibleForTesting
+    String normalizeInstanceName(String instanceName) {
+        return instanceName != null ? instanceName
+                : SystemConstants.FOGBOW_INSTANCE_NAME_PREFIX + getRandomUUID();
     }
 
-    private ComputeInstance getComputeInstance(GetVirtualMachineResponse.VirtualMachine vm, CloudStackUser cloudUser) {
+    @NotNull
+    @VisibleForTesting
+    ComputeInstance getComputeInstance(@NotNull GetVirtualMachineResponse.VirtualMachine vm,
```

Comment:
This method could be called _buildComputeInstance_, having the virtual machine and its disk as parameters. Invoking the _getVirtualMachineDiskSize_ method should be outside of this method.